File: 733-flood-fill/flood-fill.cpp
class Solution {
public:
    int rows;
    int cols;

    void dfs(int row, int col, vector<vector<int>>& grid, int& source, int& color) {
        if(row < 0 || row >= rows || col < 0 || col >= cols) 
            return;
        if(grid[row][col] != source) 
            return;
        
        grid[row][col] = color;
        dfs(row + 1, col, grid, source, color);
        dfs(row - 1, col, grid, source, color);
        dfs(row, col + 1, grid, source, color);
        dfs(row, col - 1, grid, source, color);
    }

    vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc, int color) {
        rows = image.size();
        cols = image[0].size();
        int source = image[sr][sc];

        if(source == color)
            return image;

        dfs(sr, sc, image, source, color);    
        return image;
    }
};

File: 733-flood-fill/flood-fill.java
class Solution {
    public int rows, cols;

    public void DFS(int i, int j, int[][] grid, int oldColor, int color) {
        if(i < 0 || i >= rows || j < 0 || j >= cols || grid[i][j] != oldColor) 
            return;
        
        grid[i][j] = color;
        DFS(i + 1, j, grid, oldColor, color);
        DFS(i - 1, j, grid, oldColor, color);
        DFS(i, j + 1, grid, oldColor, color);
        DFS(i, j - 1, grid, oldColor, color);
    }

    public int[][] floodFill(int[][] image, int sr, int sc, int color) {
        this.rows = image.length;
        this.cols = image[0].length;
        int oldColor = image[sr][sc];

        if(oldColor == color)
            return image;

        DFS(sr, sc, image, oldColor, color);
        return image;
    }
}